kafka-client/src/main/java/dev/responsive/kafka/internal/utils/StateDeserializer.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright 2024 Responsive Computing, Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package dev.responsive.kafka.internal.utils;
+
+import java.util.Optional;
+import org.apache.kafka.common.serialization.Deserializer;
+import org.apache.kafka.common.serialization.Serde;
+import org.apache.kafka.streams.state.ValueAndTimestamp;
+import org.apache.kafka.streams.state.internals.ValueAndTimestampSerde;
+
+public class StateDeserializer<K, V> {
+  private final String changelogTopic;
+  private final Deserializer<K> keyDeserializer;
+  private final Deserializer<V> valueDeserializer;
+  private final Optional<Deserializer<ValueAndTimestamp<V>>> timestampedValueDeserializer;
+
+  public StateDeserializer(
+      final boolean isTimestamped,
+      final String changelogTopic,
+      final Serde<K> keySerde,
+      final Serde<V> valueSerde
+  ) {
+    this.changelogTopic = changelogTopic;
+    this.keyDeserializer = keySerde == null ? null : keySerde.deserializer();
+    this.valueDeserializer = valueSerde == null ? null : valueSerde.deserializer();
+
+    if (isTimestamped && valueSerde != null) {
+      timestampedValueDeserializer =
+          Optional.of(new ValueAndTimestampSerde<>(valueSerde).deserializer());
+    } else {
+      timestampedValueDeserializer = Optional.empty();
+    }
+  }
+
+  public K keyFrom(final byte[] keyBytes) {
+    if (keyBytes == null || keyDeserializer == null) {
+      throw new IllegalStateException("Attempted to deserialize unprovided key");
```

Comment:
Let's add (a) which one was null (I know it should be both, but still makes it easier to debug) and (b) an actionable error message, something like "you must specify the keyDeserializer when creating the TtlProvider"

Reply:
This should indicate an internal/Responsive error, since we should never attempt to deserialize a null key/value where the key/value deserializer is null. So there is no actionable message for users, it's for us. But ack, I'll log which was null

Reply:
Just to clarify since I know we discussed some alternatives where this wouldn't be true, it's not possible to set a key-based ttlProvider without specifying the keyDeserializer. That's why this is an internal error not a user error

Reply:
yup, I did understand it was impossible! just would rather make it future proof in case anything changes. good erorr messages are always good :)